File: letterCombinationsofAPhoneNumber.java
public class Solution {
    public ArrayList<String> letterCombinations(String digits) {
        String[] digitStrings = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        ArrayList<String> letters = new ArrayList<String>();
        letters.add("");
        for (int i=0; i < digits.length(); i++){
            while (true){
                String s = letters.get(0);
                if (s.length() > i){
                    break;
                }
                letters.remove(0);
                for (char d: digitStrings[digits.charAt(i)-'0'-2].toCharArray()) {
                    letters.add(s+d);
                }
            }
        }
        return letters;
    }
}

File: letterCombinationsofAPhoneNumber.cpp
#include "leetcode.h"
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        string digitStrings[] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> letters;
        letters.push_back("");
        for (int i=0; i < digits.size(); i++){
            string digitString = digitStrings[digits[i]-'0'-2];
            while (true){
                string s = letters[0];
                if (s.size() > i){
                    break;
                }
                letters.erase(letters.begin());
                for (int j=0; j < digitString.size(); j++) {
                    letters.push_back(s + digitString[j]);
                }
            }
        }
        return letters;
    }
};
